src/llscan.h

```diff
@@ -174,36 +181,71 @@ class TypeRecord {
 
 
  private:
+  friend DetailedTypeRecord;
   std::string type_name_;
   uint64_t instance_count_;
   uint64_t total_instance_size_;
   std::set<uint64_t> instances_;
 };
 
+class DetailedTypeRecord : public TypeRecord {
+ public:
+  DetailedTypeRecord(std::string& type_name, uint64_t own_descriptors_count,
+                     uint64_t indexed_properties_count)
+      : TypeRecord(type_name),
+        own_descriptors_count_(own_descriptors_count),
+        indexed_properties_count_(indexed_properties_count) {}
+  inline uint64_t GetOwnDescriptorsCount() { return own_descriptors_count_; };
+  inline uint64_t GetIndexedPropertiesCount() {
+    return indexed_properties_count_;
+  };
+
+ private:
+  std::vector<std::string> properties_;
+  uint64_t own_descriptors_count_ = 0;
+  uint64_t indexed_properties_count_ = 0;
+};
+
 typedef std::map<std::string, TypeRecord*> TypeRecordMap;
+typedef std::map<std::string, DetailedTypeRecord*> DetailedTypeRecordMap;
 
 class FindJSObjectsVisitor : MemoryVisitor {
  public:
-  FindJSObjectsVisitor(lldb::SBTarget& target, TypeRecordMap& mapstoinstances);
+  struct MapCacheEntry {
+    std::string type_name;
+    std::vector<std::string> properties;
+    uint64_t own_descriptors_count = 0;
+    uint64_t indexed_properties_count = 0;
+    bool is_histogram;
+
+    std::unique_ptr<std::string> GetTypeNameWithProperties(
+        bool show_array_length = true, size_t max_properties = 0);
```

Comment:
Consider using an `enum ShowArrayLength { SHOW_ARRAY_LENGTH, DONT_SHOW_ARRAY_LENGTH }` for this, that's more obvious than a bool at call sites.